File: src/chrome/browser/android/preferences/cookie_controls_service_bridge.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/preferences/cookie_controls_service_bridge.h"

#include <memory>
#include "chrome/android/chrome_jni_headers/CookieControlsServiceBridge_jni.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_android.h"
#include "chrome/browser/ui/cookie_controls/cookie_controls_service.h"
#include "chrome/browser/ui/cookie_controls/cookie_controls_service_factory.h"

using base::android::JavaParamRef;

CookieControlsServiceBridge::CookieControlsServiceBridge(
    JNIEnv* env,
    const JavaParamRef<jobject>& obj,
    const base::android::JavaParamRef<jobject>& jprofile)
    : jobject_(obj) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(jprofile);
  service_ = CookieControlsServiceFactory::GetForProfile(profile);
  service_->AddObserver(this);
}

void CookieControlsServiceBridge::HandleCookieControlsToggleChanged(
    JNIEnv* env,
    jboolean checked) {
  service_->HandleCookieControlsToggleChanged(checked);
}

void CookieControlsServiceBridge::SendCookieControlsUIChanges() {
  bool checked = service_->GetToggleCheckedValue();
  bool enforced = service_->ShouldEnforceCookieControls();
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_CookieControlsServiceBridge_sendCookieControlsUIChanges(
      env, jobject_, checked, enforced);
}

void CookieControlsServiceBridge::OnThirdPartyCookieBlockingPrefChanged() {
  SendCookieControlsUIChanges();
}

void CookieControlsServiceBridge::OnThirdPartyCookieBlockingPolicyChanged() {
  SendCookieControlsUIChanges();
}

CookieControlsServiceBridge::~CookieControlsServiceBridge() = default;

void CookieControlsServiceBridge::Destroy(JNIEnv* env,
                                          const JavaParamRef<jobject>& obj) {
  delete this;
}

static jlong JNI_CookieControlsServiceBridge_Init(
    JNIEnv* env,
    const JavaParamRef<jobject>& obj,
    const base::android::JavaParamRef<jobject>& jprofile) {
  return reinterpret_cast<intptr_t>(
      new CookieControlsServiceBridge(env, obj, jprofile));
}

File: src/chrome/android/java/src/org/chromium/chrome/browser/settings/website/CookieControlsServiceBridge.java
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.settings.website;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.profiles.Profile;

/**
 * Communicates between CookieControlsService (C++ backend) and observers in the Incognito NTP Java
 * UI.
 */
public class CookieControlsServiceBridge {
    /**
     * Interface for a class that wants to receive cookie controls updates from
     * CookieControlsServiceBridge.
     */
    public interface CookieControlsServiceObserver {
        /**
         * Called when there is an update in the cookie controls that should be reflected in the UI.
         * @param checked A boolean indicating whether the toggle indicating third-party cookies are
         *         currently being blocked should be checked or not.
         * @param enforced A boolean indicating if third-party cookies being blocked is currently
         *         enforced by policy/cookie settings or not.
         */
        public void sendCookieControlsUIChanges(boolean checked, boolean enforced);
    }

    private long mNativeCookieControlsServiceBridge;
    private CookieControlsServiceObserver mObserver;

    /**
     * Initializes a CookieControlsServiceBridge instance.
     * @param observer An observer to call with updates from the cookie controls service.
     * @param profile The Profile instance to observe.
     */
    public CookieControlsServiceBridge(CookieControlsServiceObserver observer, Profile profile) {
        mObserver = observer;
        mNativeCookieControlsServiceBridge = CookieControlsServiceBridgeJni.get().init(
                CookieControlsServiceBridge.this, profile);
    }

    /**
     * Destroys the native counterpart of this class.
     */
    public void destroy() {
        if (mNativeCookieControlsServiceBridge != 0) {
            CookieControlsServiceBridgeJni.get().destroy(
                    mNativeCookieControlsServiceBridge, CookieControlsServiceBridge.this);
            mNativeCookieControlsServiceBridge = 0;
        }
    }

    /**
     * Updates the CookieControlsService on the status of the toggle, and thus the state of
     * third-party cookie blocking in incognito.
     * @param enable A boolean indicating whether the toggle has been switched on or off.
     */
    public void handleCookieControlsToggleChanged(boolean enable) {
        CookieControlsServiceBridgeJni.get().handleCookieControlsToggleChanged(
                mNativeCookieControlsServiceBridge, enable);
    }

    @CalledByNative
    private void sendCookieControlsUIChanges(boolean checked, boolean enforced) {
        mObserver.sendCookieControlsUIChanges(checked, enforced);
    }

    @NativeMethods
    interface Natives {
        long init(CookieControlsServiceBridge caller, Profile profile);
        void destroy(long nativeCookieControlsServiceBridge, CookieControlsServiceBridge caller);
        void handleCookieControlsToggleChanged(
                long nativeCookieControlsServiceBridge, boolean enable);
    }
}
